sklearnex/tests/test_patching.py

```diff
@@ -116,7 +118,65 @@ def _load_all_models(patched):
 UNPATCHED_MODELS = _load_all_models(patched=False)
 
 
-@pytest.mark.parametrize(("patched", "unpatched"), zip(PATCHED_MODELS, UNPATCHED_MODELS))
-def test_is_patched_instance(patched, unpatched):
+@pytest.mark.parametrize("name", PATCHED_MODELS.keys())
+def test_is_patched_instance(name):
+    patched = PATCHED_MODELS[name]()
+    unpatched = UNPATCHED_MODELS[name]()
     assert is_patched_instance(patched), f"{patched} is a patched instance"
     assert not is_patched_instance(unpatched), f"{unpatched} is an unpatched instance"
+
+
+@pytest.mark.parametrize("name", PATCHED_MODELS.keys())
+def test_if_control_n_jobs_decorated(name):
+    patched = PATCHED_MODELS[name]
+    text, line = inspect.getsourcelines(patched)
+    loc = inspect.getfile(patched)
+
+    # Inconsistency in getsource for python3.8 requires additional checks
+    while len(text[0].strip()) != 0 and line > 0:
+        line -= 1
+        text.insert(0, linecache.getline(loc, line))
+
+    # Get text before class definition
+    text = "".join(text).split("class " + name + "(")[0]
+
+    lines = "".join(re.findall(r"@\S+", text))
+    assert (
+        "@control_n_jobs" in lines
+    ), f"'{name}' does not support sklearnex n_job design rules"
+
+
+@pytest.mark.parametrize(
+    "name", [i for i in PATCHED_MODELS if "daal4py" in PATCHED_MODELS[i].__module__]
+)
+def test_daal4py_run_with_n_jobs_decorated(name):
+    patched = PATCHED_MODELS[name]
+    funcs = [i for i in dir(patched) if inspect.isfunction(getattr(patched, i))]
+    for i in funcs:
+        text = inspect.getsource(getattr(patched, i))
+
+        # find if it contains @run_with_n_jobs
+        njobs = "@run_with_n_jobs" in text
+        # find if it contains a call to a daal4py function
+        onedal_method = len(re.findall(r"daal4py\.\S+\(", text)) > 0
```

Comment:
`@run_with_n_jobs`-decorated function may contain daal4py call or nested calls in other function like it's done in TSNE/KMeans. Therefore, runtime code analysis is not usable for this check.

Reply:
daal4py test has been removed.